runtime/parachains/src/paras_inherent.rs

```diff
@@ -149,122 +279,190 @@ pub mod pallet {
 		}
 	}
 
+	/// Collect all freed cores based on storage data.
+	///
+	/// The parameter `freed_concluded` contains all core indicies that became
+	/// free due to candidate that became available.
+	pub(crate) fn collect_all_freed_cores<T, I>(
+		freed_concluded: I,
+	) -> BTreeMap<CoreIndex, FreedReason>
+	where
+		I: core::iter::IntoIterator<Item = (CoreIndex, CandidateHash)>,
+		T: Config,
+	{
+		// Handle timeouts for any availability core work.
+		let availability_pred = <scheduler::Pallet<T>>::availability_timeout_predicate();
+		let freed_timeout = if let Some(pred) = availability_pred {
+			<inclusion::Pallet<T>>::collect_pending(pred)
+		} else {
+			Vec::new()
+		};
+
+		// Schedule paras again, given freed cores, and reasons for freeing.
+		let freed = freed_concluded
+			.into_iter()
+			.map(|(c, _hash)| (c, FreedReason::Concluded))
+			.chain(freed_timeout.into_iter().map(|c| (c, FreedReason::TimedOut)))
+			.collect::<BTreeMap<CoreIndex, FreedReason>>();
+		freed
+	}
+
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
 		/// Enter the paras inherent. This will process bitfields and backed candidates.
 		#[pallet::weight((
-			MINIMAL_INCLUSION_INHERENT_WEIGHT + data.backed_candidates.len() as Weight * BACKED_CANDIDATE_WEIGHT,
+			paras_inherent_total_weight::<T>(
+				data.backed_candidates.as_slice(),
+				data.bitfields.as_slice(),
+				data.disputes.as_slice(),
+			),
 			DispatchClass::Mandatory,
 		))]
 		pub fn enter(
 			origin: OriginFor<T>,
 			data: ParachainsInherentData<T::Header>,
 		) -> DispatchResultWithPostInfo {
+			ensure_none(origin)?;
+
+			ensure!(!Included::<T>::exists(), Error::<T>::TooManyInclusionInherents);
+			Included::<T>::set(Some(()));
+
 			let ParachainsInherentData {
-				bitfields: signed_bitfields,
-				backed_candidates,
+				bitfields: mut signed_bitfields,
+				mut backed_candidates,
 				parent_header,
-				disputes,
+				mut disputes,
 			} = data;
 
-			ensure_none(origin)?;
-			ensure!(!Included::<T>::exists(), Error::<T>::TooManyInclusionInherents);
-
-			// Check that the submitted parent header indeed corresponds to the previous block hash.
 			let parent_hash = <frame_system::Pallet<T>>::parent_hash();
+			// Check that the submitted parent header indeed corresponds to the previous block hash.
 			ensure!(
 				parent_header.hash().as_ref() == parent_hash.as_ref(),
```

Comment:
This is one of the basics and will remain checked in the provisioningsubsystem. This would indicate a bug rather anything else iiuc. So in reference to the above statement, this is fine.